api/tests/mocha/controllers/contact.spec.js

```diff
@@ -0,0 +1,402 @@
+const sinon = require('sinon');
+const auth = require('../../../src/auth');
+const dataContext = require('../../../src/services/data-context');
+const serverUtils = require('../../../src/server-utils');
+const { Contact, Qualifier, InvalidArgumentError} = require('@medic/cht-datasource');
+const {expect} = require('chai');
+const controller = require('../../../src/controllers/contact');
+
+describe('Contact Controller', () => {
+  const userCtx = { hello: 'world' };
+  let getUserCtx;
+  let isOnlineOnly;
+  let hasAllPermissions;
+  let dataContextBind;
+  let serverUtilsError;
+  let req;
+  let res;
+
+  beforeEach(() => {
+    getUserCtx = sinon
+      .stub(auth, 'getUserCtx')
+      .resolves(userCtx);
+    isOnlineOnly = sinon.stub(auth, 'isOnlineOnly');
+    hasAllPermissions = sinon.stub(auth, 'hasAllPermissions');
+    dataContextBind = sinon.stub(dataContext, 'bind');
+    serverUtilsError = sinon.stub(serverUtils, 'error');
+    res = {
+      json: sinon.stub(),
+    };
+  });
+
+  afterEach(() => sinon.restore());
+
+  describe('v1', () => {
+    describe('get', () => {
+      let contactGet;
+      let contactGetWithLineage;
+
+      beforeEach(() => {
+        req = {
+          params: { uuid: 'uuid' },
+          query: { }
+        };
+        contactGet = sinon.stub();
+        contactGetWithLineage = sinon.stub();
+        dataContextBind
+          .withArgs(Contact.v1.get)
+          .returns(contactGet);
+        dataContextBind
+          .withArgs(Contact.v1.getWithLineage)
+          .returns(contactGetWithLineage);
+      });
+
+      afterEach(() => {
+        expect(getUserCtx.calledOnceWithExactly(req)).to.be.true;
+        expect(isOnlineOnly.calledOnceWithExactly(userCtx)).to.be.true;
```

Comment:
> What if afterEach precedence gets messed up and we end up calling another afterEach first that does a bunch of cleanup.

If the other afterEach is pretty messy or too dependent then this would mean that these assertions would fail, right? Which in turn would make us refactor this implementation.

Reply:
Agree to disagree. I think it's really confusing to have assertions in afterEach. How does it even look when your afterEach assertions fail? Which test ends up reported as failing? Do coverage tools also consider assertions in afterEach blocks?

Why would we not follow best practices in all cases and qualify our case as "special" (or frictionless)? 

I'm guilty of also adding assertions once in an afterEach :(

Reply:
> Why would we not follow best practices in all cases 

I guess I see this as a case of competing best practices that we have to find the balance between....  On one hand it is great to have the assertions in the `afterEach` since it guarantees that they get run for each test (you do not have to rely on the engineer remembering to include them in any new tests) and it reduces duplicated code.  On the other hand, as you mentioned, it makes it slightly less clear what all is being asserted in the tests and it does mess with the Mocha run report.  IIRC, a failure of an assertion in the `afterEach` will result Mocha marking the tests as "passed", but failing the suite (and providing a helpful message indicating the afterEach failed).  

Either way, this is not a hill I want to die on, but we are already doing assertions like this throughout our existing automated tests.  So, if this is a practice we do not want to continue with, we should capture that in our docs (maybe in the [automated tests style guide](https://docs.communityhealthtoolkit.org/contribute/code/core/style-guide-automated-e2e-tests/)).

Reply:
The argument that we are already doing something wrong so it's fine to continue doesn't really seem valid for me. 
I don't think this is such a pain to implement _here_ and I don't think this change request is outrageous in any way. Yet you are seeking ground for dismissal. 
I simply don't understand.

Reply:
> if this is a practice we do not want to continue with, we should capture that in our docs (maybe in the [automated tests style guide](https://docs.communityhealthtoolkit.org/contribute/code/core/style-guide-automated-e2e-tests/)).

I am only seeking clarity in our approach for future reference (so we do not have to re-hash this discussion on a future PR).  This is something that we already do in our code base. If we don't want to do it anymore, lets document that and move on.

Reply:
Sure, let's document this.